src/platform/backends/hyperv_api/virtdisk/virtdisk_snapshot.cpp

```diff
@@ -0,0 +1,284 @@
+    }();
+
+    const bool should_merge_head = (self_path == head_parent) && vm.get_num_snapshots() == 1;
+    const bool should_reparent_head = (self_path == head_parent) && vm.get_num_snapshots() > 1;
+
+    // If the head's parent is this, and this is the last snapshot,
+    // we need to merge the head to the snapshot first.
+    if (should_merge_head)
+    {
+        // Head is attached to the snapshot that's going to be deleted.
+        if (const auto merge_r = VirtDisk().merge_virtual_disk_to_parent(head_path); merge_r)
+        {
+            mpl::debug(log_category,
+                       "Successfully merged head differencing disk `{}` to parent disk `{}`",
+                       head_path,
+                       self_path);
+            mpl::debug(log_category, "Removing the merged head disk file: `{}`", head_path);
+            MP_FILEOPS.remove(head_path);
+        }
+        else
+        {
+            throw CreateVirtdiskSnapshotError{
+                merge_r,
+                "Could not merge head differencing disk to the edge snapshot"};
+        }
+    }
+
+    //  1: Merge this to its parent
+    if (const auto merge_r = VirtDisk().merge_virtual_disk_to_parent(self_path); merge_r)
+    {
+        const auto& parent_path = parent ? make_snapshot_path(*parent) : base_vhdx_path;
+        mpl::debug(log_category,
+                   "Successfully merged differencing disk `{}` to parent disk `{}`",
+                   self_path,
+                   parent_path);
+
+        // The actual reparenting of the children needs to happen here.
+        // Reparenting is not a simple "-> now this is your parent" like thing. The children
+        // include parent's metadata calculated based on actual contents, and merging a child disk
+        // to parent updates its parent's metadata, too.
+        // Hence, the reparenting operation not only needs to happen to the orphaned children,
+        // but also to the existing children of the parent as well, so the updated metadata of the
+        // parent could be reflected to the all.
+        const auto children_to_reparent =
+            vm.view_snapshots([&parent, this_index = this->get_index()](const Snapshot& ss) {
+                return
+                    // Exclude self.
+                    (ss.get_index() != this_index) &&
+                    // set_parent() for the orphaned children happens before erase() call
+                    // so they're already adopted by the self's parent at this point.
+                    (parent ? (ss.get_parents_index() == parent->get_index())
+                            : ss.get_parents_index() == 0);
+            });
+        reparent_snapshot_disks(children_to_reparent, parent_path);
+
+        // Determine if head needs reparenting
+        if (should_reparent_head)
+        {
+            if (const auto hr = VirtDisk().reparent_virtual_disk(head_path, parent_path); hr)
```

Comment:
`hr` is a little cryptic here. How about `res` instead?

Reply:
Done